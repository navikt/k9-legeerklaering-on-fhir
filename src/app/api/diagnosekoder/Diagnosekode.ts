export interface Diagnosekode {
    readonly code: string;
    readonly text: string;
}

export const instanceOfDiagnosekode = (obj: any): obj is Diagnosekode =>
    'code' in obj && 'text' in obj &&
    typeof obj.code === 'string' &&
    typeof obj.text === 'string';

export const instanceOfDiagnosekodeArray = (arr: any): arr is Diagnosekode[] => arr instanceof Array && (
    arr.length === 0 || instanceOfDiagnosekode(arr[0])
)
