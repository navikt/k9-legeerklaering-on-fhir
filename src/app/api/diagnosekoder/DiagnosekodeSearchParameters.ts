export interface DiagnosekodeSearchParameters {
    readonly searchText: string;
    readonly pageNumber: number;
}

const search = "search";
const page = "page";

export const searchParametersToUrl = (baseURL: URL, params: DiagnosekodeSearchParameters): URL => {
    const url = new URL(baseURL);
    if(params.searchText.length > 0) {
        url.searchParams.append(search, params.searchText)
    }
    url.searchParams.append(page, params.pageNumber.toString());
    return url;
}

export const searchParametersFromUrl = (url: URL): DiagnosekodeSearchParameters => {
    const searchText = url.searchParams.get(search) || '';
    const pageNumber = Number(url.searchParams.get(page)) || 1;
    return {searchText, pageNumber};
}
