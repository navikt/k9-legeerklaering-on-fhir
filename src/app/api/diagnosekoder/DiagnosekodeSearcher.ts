import type {Diagnosekode} from "@/app/api/diagnosekoder/Diagnosekode";
import type {DiagnosekodeSearchResult} from "@/app/api/diagnosekoder/DiagnosekodeSearchResult";

/**
 * We add an uppercased text for doing case-insensitive search.
 */
export interface DiagnosekodeWithUppercased extends Diagnosekode {
    readonly uppercasedCode: string;
    readonly uppercasedText: string;
}


/**
 * Helper class that provides case-insensitive search of given diagnosekoder array.
 */
export class DiagnosekodeSearcher {
    public readonly diagnosekoderWithUppercased: DiagnosekodeWithUppercased[];

    public constructor(
        public readonly diagnosekoder: Diagnosekode[],
        public readonly pageSize: number,
    ) {
        this.diagnosekoderWithUppercased = this.diagnosekoder.map(dk =>
            ({...dk, uppercasedCode: dk.code.toUpperCase(), uppercasedText: dk.text.toUpperCase()})
        );
    }

    /**
     * Case-insensitive search of diagnosekoder with given searchText.
     *
     * This returns diagnosekode entries that includes searchText either in code or text props.
     *
     * If searchText is empty, the first <pageSize> number of diagnosekoder is returned.
     * @param searchText
     * @param pageNumber
     */
    public search(searchText: string, pageNumber: number): DiagnosekodeSearchResult {
        const sliceStart = (pageNumber - 1) * this.pageSize;
        const sliceEnd = sliceStart + this.pageSize;
        if (searchText.length === 0) {
            return {
                diagnosekoder: this.diagnosekoder.slice(sliceStart, sliceEnd),
                pageNumber,
                hasMore: this.diagnosekoder.length > pageNumber * this.pageSize,
            };
        } else {
            const searchTextUppercased = searchText.trim().toUpperCase()
            const found = this.diagnosekoderWithUppercased.filter(diagnosekode => {
                return  diagnosekode.uppercasedCode.includes(searchTextUppercased) || diagnosekode.uppercasedText.includes(searchTextUppercased)
            }).map(dk => ({code: dk.code, text: dk.text}));
            return {
                diagnosekoder: found.slice(sliceStart, sliceEnd),
                pageNumber: pageNumber,
                hasMore: found.length > pageNumber * this.pageSize,
            };
        }
    }
}