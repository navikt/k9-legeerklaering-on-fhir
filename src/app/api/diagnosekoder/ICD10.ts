import ICD10 from './ICD10.json';

export interface Diagnosekode {
    readonly code: string;
    readonly text: string;
}

/**
 * We add an uppercased text for doing case-insensitive search.
 */
export interface DiagnosekodeWithUppercased extends Diagnosekode {
    readonly uppercasedCode: string;
    readonly uppercasedText: string;
}

export const instanceOfDiagnosekode = (obj: any): obj is Diagnosekode => 'code' in obj && 'text' in obj;

export const instanceOfDiagnosekodeArray = (arr: any): arr is Diagnosekode[] => arr instanceof Array && (
    arr.length === 0 || instanceOfDiagnosekode(arr[0])
)

export const diagnosekoder: Diagnosekode[] = ICD10;

export const diagnosekoderWithUppercased: DiagnosekodeWithUppercased[] =
    diagnosekoder.map(dk => ({...dk, uppercasedCode: dk.code.toUpperCase(), uppercasedText: dk.text.toUpperCase()}))

export const searchDiagnosekoder = (searchText: string): Diagnosekode[] => {
    const searchTextUppercased = searchText.trim().toUpperCase()
    return diagnosekoderWithUppercased.filter(diagnosekode => {
        return  diagnosekode.uppercasedCode.includes(searchTextUppercased) || diagnosekode.uppercasedText.includes(searchTextUppercased)
    }).map(dk => ({code: dk.code, text: dk.text}));
}
