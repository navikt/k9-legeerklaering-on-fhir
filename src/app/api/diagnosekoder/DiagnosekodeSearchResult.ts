import Diagnosekode, {instanceOfDiagnosekode} from "@navikt/diagnosekoder/Diagnosekode";

export interface DiagnosekodeSearchResult {
    readonly diagnosekoder: Diagnosekode[];
    readonly pageNumber: number;
    readonly hasMore: boolean;
}

const instanceOfDiagnosekodeArray = (arr: any): arr is Diagnosekode[] =>
    arr instanceof Array && (arr.length === 0 || instanceOfDiagnosekode(arr[0]))

export const instanceOfDiagnosekodeSearchResult = (some: any): some is DiagnosekodeSearchResult =>
    'diagnosekoder' in some &&
    instanceOfDiagnosekodeArray(some.diagnosekoder) &&
    'hasMore' in some
