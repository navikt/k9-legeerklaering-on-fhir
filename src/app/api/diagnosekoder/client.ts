import validateRoute from "@/utils/validateRoute";
import {
    DiagnosekodeSearchResult,
    instanceOfDiagnosekodeSearchResult
} from "@/app/api/diagnosekoder/DiagnosekodeSearchResult";
import {searchParametersToUrl} from "@/app/api/diagnosekoder/DiagnosekodeSearchParameters";


export const diagnosekoderApiPath = validateRoute('/api/diagnosekoder');


export const searchDiagnosekoderFetch = async (searchText: string, pageNumber: number, abortSignal: AbortSignal): Promise<DiagnosekodeSearchResult> => {
    const baseUrl = new URL(window.location.origin);
    const url = searchParametersToUrl(
        new URL(diagnosekoderApiPath, baseUrl),
        {searchText, pageNumber}
    );
    const resp = await fetch(url.toString(), {signal: abortSignal})
    if (resp.ok) {
        const data = await resp.json()
        if (instanceOfDiagnosekodeSearchResult(data)) {
            return data;
        } else {
            const someData = JSON.stringify(data).substring(0, 80)
            throw new Error(`json returned from fetch call did not match expected DiagnosekodeSearchResult interface. (${someData})`)
        }
    } else {
        throw new Error(`Unexpected status code when fetching diagnosekoder (${resp.status} - ${resp.statusText})`)
    }
}


