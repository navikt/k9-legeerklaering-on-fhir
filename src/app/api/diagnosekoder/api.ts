import validateRoute from "@/utils/validateRoute";
import type {Diagnosekode} from "@/app/api/diagnosekoder/ICD10";
import {instanceOfDiagnosekodeArray} from "@/app/api/diagnosekoder/ICD10";

export const diagnosekoderApiPath = validateRoute('/api/diagnosekoder');
export const pageSize = 100;

export interface DiagnosekodeSearchResult {
    readonly diagnosekoder: Diagnosekode[];
    readonly pageNumber: number;
    readonly foundCount: number;
}

export const instanceOfDiagnosekodeSearchResult = (some: any): some is DiagnosekodeSearchResult =>
    'diagnosekoder' in some &&
    instanceOfDiagnosekodeArray(some.diagnosekoder) &&
    'foundCount' in some

export const searchDiagnosekoderFetch = async (searchText: string, pageNumber: number, abortSignal: AbortSignal): Promise<DiagnosekodeSearchResult> => {
    const url = new URL(diagnosekoderApiPath, window.location.origin)
    if(searchText.length > 0) {
        url.searchParams.append("search", searchText)
    }
    url.searchParams.append("page", pageNumber.toString());
    const resp = await fetch(url.toString(), {signal: abortSignal})
    if (resp.ok) {
        const data = await resp.json()
        if (instanceOfDiagnosekodeSearchResult(data)) {
            return data;
        } else {
            throw new Error(`json returned from fetch call did not match expected DiagnosekodeSearchResult interface. (${data.substring(0, 30)})`)
        }
    } else {
        throw new Error(`Unexpected status code when fetching diagnosekoder (${resp.status} - ${resp.statusText})`)
    }
}