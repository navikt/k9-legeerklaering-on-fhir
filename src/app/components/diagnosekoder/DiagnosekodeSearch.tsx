"use client"

import {Alert, Button, Search, Table} from "@navikt/ds-react";
import {Diagnosekode} from "@/app/api/diagnosekoder/ICD10";
import {
    ForwardedRef,
    forwardRef,
    KeyboardEvent,
    useCallback,
    useEffect,
    useImperativeHandle,
    useRef,
    useState
} from "react";
import {DiagnosekodeSearchResult, pageSize, searchDiagnosekoderFetch} from "@/app/api/diagnosekoder/api";
import {ArrowRightIcon, ChevronDownDoubleIcon} from '@navikt/aksel-icons';

import diagnosekoderCss from './diagnosekoder.module.css';

interface DiagnosekoderProp {
    readonly diagnosekoder: Diagnosekode[];
}

export interface OnSelectedDiagnose {
    readonly onSelectedDiagnose?: (diagnosekode: Diagnosekode) => void;
}

const noop = () => void 0;

const DiagnosekodeRow = ({code, text, onSelectedDiagnose}: Diagnosekode & OnSelectedDiagnose) => {
    const handleSelected = onSelectedDiagnose ?
            () => onSelectedDiagnose({code, text}) :
            noop;
    // Parent component DiagnosekodeTable depends somewhat on the html structure here, by using css queries.
    // Take care not to change this html without looking at the DiagnosekodeTable css queries.
    return (
        <Table.Row onClick={(ev) => {ev.stopPropagation(); handleSelected()}} className={diagnosekoderCss.diagnosekoderow}>
            <Table.DataCell>{code}</Table.DataCell>
            <Table.DataCell>{text}</Table.DataCell>
            <Table.DataCell>
                <Button
                    onClick={(ev) => {ev.stopPropagation(); handleSelected()}}
                    variant="tertiary"
                    size="small"
                    aria-label="Velg"
                    icon={<ArrowRightIcon aria-hidden />}>
                </Button>
            </Table.DataCell>
        </Table.Row>
    )
}

const DiagnosekodeRows = ({diagnosekoder, onSelectedDiagnose}: DiagnosekoderProp & OnSelectedDiagnose) => {
    return diagnosekoder.map(diagnosekode => <DiagnosekodeRow key={diagnosekode.code} {...diagnosekode} onSelectedDiagnose={onSelectedDiagnose} />)
}

const FetchDiagnosekoderErrorAlert = ({error}: {error: Error}) => {
    return (
        <Alert className="my-4" variant="error">
            <p>Søk etter diagnosekode feilet. Prøv igjen.</p>
            <p>
                <small>{error.message}</small>
            </p>
        </Alert>
    )
}

export interface DiagnosekodeTableProps extends DiagnosekoderProp, OnSelectedDiagnose {
    onKeyDown?(event: KeyboardEvent<HTMLTableElement>): void;
}

export interface DiagnosekodeTableMethods {
    /**
     * Used by parent component when it wants to focus the "select" button of the first row in the search result table.
     */
    focusFirst(): void;

    /**
     * Used by parent component when it wants to focus the "select" button of the next row in the search result table.
     *
     * This assumes that a "select" button in the table currently has focus. If no button has focus, it will call focusFirst().
     */
    focusNext(): void;

    /**
     * Used by parent component when it wants to focus the "select" button of previous row in the search result table.
     *
     * If there is no focuse row, or no row above currently focused row, nothing is done.
     */
    focusPrevious(): void;
}

const DiagnosekodeTable = forwardRef(({diagnosekoder, onSelectedDiagnose, onKeyDown}: DiagnosekodeTableProps, ref: ForwardedRef<DiagnosekodeTableMethods>) => {
    // The ref and css selector stuff here depends on the html structure of the subcomponents used (DiagnosekodeRows, ...).
    // So we must take care not to break this. Could probably avoid the fragility by pushing using more refs and imperative
    // handlers further down the component chain, but don't want to spend time doing that atm.
    const tableRef = useRef<HTMLTableElement>(null);
    useImperativeHandle(ref, () => {
        const rowWithFocusedButton = () => tableRef.current?.querySelector("tr:has(button:focus)");
        const noButtonHasFocus = () => rowWithFocusedButton() === null;
        return {
            focusFirst() {
                const button = tableRef.current?.querySelector("button");
                button?.focus();
            },
            focusNext() {
                const nextButton = tableRef.current?.querySelector<HTMLButtonElement>("tr:has(button:focus) + tr button");
                if (nextButton) {
                    nextButton.focus();
                } else if(noButtonHasFocus()) {
                    this.focusFirst();
                }
            },
            focusPrevious() {
                const focusedRow = rowWithFocusedButton();
                const prevRowButton = focusedRow?.previousElementSibling?.querySelector("button");
                prevRowButton?.focus();
            }
        }
    }, [tableRef]);

    return (
        <Table ref={tableRef} onKeyDown={onKeyDown}>
            <Table.Header>
                <Table.Row>
                    <Table.HeaderCell scope="col">Kode</Table.HeaderCell>
                    <Table.HeaderCell scope="col">Beskrivelse</Table.HeaderCell>
                    <Table.HeaderCell scope="col"></Table.HeaderCell>
                </Table.Row>
            </Table.Header>
            <Table.Body>
                <DiagnosekodeRows diagnosekoder={diagnosekoder} onSelectedDiagnose={onSelectedDiagnose} />
            </Table.Body>
        </Table>
    )
})
DiagnosekodeTable.displayName = "DiagnosekodeTable"

const emptySearchResult: DiagnosekodeSearchResult = {
    diagnosekoder: [],
    pageNumber: 0,
    foundCount: 0
};

interface SearchParams {
    readonly searchText: string;
    readonly pageNumber: number;
}

const initSearchParams: SearchParams = {
    searchText: "",
    pageNumber: 1,
}

const keycodes = {
    ArrowDown: 'ArrowDown',
    ArrowUp: 'ArrowUp',
}

const DiagnosekodeSearch = ({onSelectedDiagnose}: OnSelectedDiagnose) => {
    const [searchParams, setSearchParams] = useState(initSearchParams)
    const [searchResult, setSearchResult] = useState<DiagnosekodeSearchResult>(emptySearchResult)
    const [error, setError] = useState<Error | null>(null)
    const diagnosekodeTableRef = useRef<DiagnosekodeTableMethods>(null)

    const trimmedSearchText = searchParams.searchText.trim(); // No point in doing a new search if user just adds/removes a space
    const pageNumber = searchParams.pageNumber;

    useEffect(() => {
        let isSubscribed = true;
        // TODO AbortController for the fetch?
        const fetchDiagnosekoder = async (): Promise<void> => {
            const searchResult = await searchDiagnosekoderFetch(trimmedSearchText, pageNumber);
            if (isSubscribed) {
                if (pageNumber > 1) {
                    // The search result is a paging continuation, add it to the existing result
                    setSearchResult(existing =>
                        ({
                            diagnosekoder: [...existing.diagnosekoder, ...searchResult.diagnosekoder],
                            pageNumber: searchResult.pageNumber,
                            foundCount: searchResult.foundCount,
                        })
                    );
                } else { // Search result is a new first page result
                    setSearchResult(searchResult)
                }
            }// else another fetch has been started, ignore this
        };

        fetchDiagnosekoder()
            .then(
                () => {
                    setError(null)
                } , err => {
                    if(err instanceof Error) {
                        setError(err)
                    } else {
                        setError(new Error(err))
                    }
                }
            )
        return () => {
            isSubscribed = false;
        };
    }, [trimmedSearchText, pageNumber])

    const handleSearchChanged = useCallback((searchText: string) => setSearchParams({searchText, pageNumber: initSearchParams.pageNumber}), [])
    const handleShowMoreClicked = useCallback(() => setSearchParams({searchText: trimmedSearchText, pageNumber: pageNumber + 1}), [trimmedSearchText, pageNumber])
    const handleKeyUp = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter") {
            if (searchResult.diagnosekoder.length === 1 && onSelectedDiagnose !== undefined) {
                onSelectedDiagnose(searchResult.diagnosekoder[0]);
            }
        } else if (event.code === keycodes.ArrowDown && searchResult.diagnosekoder.length > 0) {
            // Focus the first choose button in search result list
            diagnosekodeTableRef.current?.focusFirst();
        }
    }, [onSelectedDiagnose, searchResult])

    const handleDiagnosekodeTableKeyDown = (ev: KeyboardEvent<HTMLTableElement>) => {
        if (ev.code === keycodes.ArrowDown) { // arrow down was pressed while in the table, focus the next select button, if possible.
            diagnosekodeTableRef.current?.focusNext();
        } else if (ev.code === keycodes.ArrowUp) {
            diagnosekodeTableRef.current?.focusPrevious();
        }
    }

    const resultElements = error ?
        <FetchDiagnosekoderErrorAlert error={error} /> :
        <DiagnosekodeTable
            ref={diagnosekodeTableRef}
            diagnosekoder={searchResult.diagnosekoder}
            onSelectedDiagnose={onSelectedDiagnose}
            onKeyDown={handleDiagnosekodeTableKeyDown}
        />;

    const showMoreElements = searchResult.foundCount > searchResult.pageNumber * pageSize ?
        <div className="flex justify-center py-4">
            <Button onClick={handleShowMoreClicked} variant="tertiary" icon={<ChevronDownDoubleIcon aria-hidden />}>Vis flere rader</Button>
        </div> :
        null;

    return (
        <>
                <Search
                    autoFocus
                    label="Søk med kode eller beskrivelse"
                    onChange={handleSearchChanged}
                    onKeyUp={handleKeyUp}
                    variant="simple"
                    clearButton={false}
                    className={diagnosekoderCss.searchinput}
                    placeholder="Søk (kode eller beskrivelse)"
                />
                <div className={diagnosekoderCss.searchresult}>
                    {resultElements}
                    {showMoreElements}
                </div>
        </>
    )
}

export default DiagnosekodeSearch;