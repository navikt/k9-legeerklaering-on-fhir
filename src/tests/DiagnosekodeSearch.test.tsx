import {render, waitFor, screen} from "@testing-library/react";
import DiagnosekodeSearch from "@/app/components/diagnosekoder/DiagnosekodeSearch";
import {type Diagnosekode} from "@navikt/diagnosekoder";
import {searchDiagnosekoderFetch} from "@/app/api/diagnosekoder/client";
import {fakeDiagnosekoder} from "@/tests/DiagnosekodeSearcher.test";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

jest.mock("@/app/api/diagnosekoder/client")

const mockedSearchDiagnosekoderFetch = jest.mocked(searchDiagnosekoderFetch);

describe("DiagnosekodeSearch works", () => {
    test("initial empty search returns pages of results", async () => {
        const firstPageDiagnosekoder = fakeDiagnosekoder.slice(0, 3);
        const secondPageDiagnosekoder = fakeDiagnosekoder.slice(3, 100);
        // This mocks a first result with no search input
        mockedSearchDiagnosekoderFetch.mockResolvedValue({
            diagnosekoder: firstPageDiagnosekoder,
            hasMore: true,
            pageNumber: 1,
        });
        const rendered = render(<DiagnosekodeSearch />)

        // We must wait until the (mocked) fetch has returned to see the table rendered with the result
        await waitFor(() => {
            expect(mockedSearchDiagnosekoderFetch).toHaveReturned()
        })

        const table = await rendered.findByRole("table")
        expect(table).toMatchSnapshot()

        for(const dk of firstPageDiagnosekoder) {
            expect(await rendered.findByText(dk.code)).toBeInTheDocument()
            expect(await rendered.findByText(dk.text)).toBeInTheDocument()
        }
        for(const dk of secondPageDiagnosekoder) {
            expect(rendered.queryByText(dk.code)).toBeNull()
            expect(rendered.queryByText(dk.code)).not.toBeInTheDocument()
            expect(rendered.queryByText(dk.text)).not.toBeInTheDocument()
        }

        // Setup mock to return next page of results
        mockedSearchDiagnosekoderFetch.mockReset().mockResolvedValue({
            diagnosekoder: secondPageDiagnosekoder,
            hasMore: false,
            pageNumber: 2,
        })

        const getMoreBtn = screen.getByText(/Vis flere rader/i);
        await userEvent.click(getMoreBtn);

        // Wait for fetch of next page to return
        await waitFor(() => {
            expect(mockedSearchDiagnosekoderFetch).toHaveReturned()
        })

        // All diagnosekode rows should be in the document now
        for(const dk of fakeDiagnosekoder) {
            expect(rendered.getByText(dk.code)).toBeInTheDocument()
            expect(rendered.getByText(dk.text)).toBeInTheDocument()
        }

        // Show more rows should not display anymore
        expect(getMoreBtn).not.toBeInTheDocument()
    })

    test("searching for a diagnose should work", async () => {
        mockedSearchDiagnosekoderFetch.mockReset().mockResolvedValue({
            diagnosekoder: fakeDiagnosekoder.slice(0,1),
            hasMore: false,
            pageNumber: 1,
        })
        const rendered = render(<DiagnosekodeSearch />)
        const searchInput = rendered.getByRole("searchbox");
        expect(searchInput).toBeInTheDocument()
        const inputCode = "a001"
        await userEvent.type(searchInput, inputCode);
        await waitFor(() => expect(mockedSearchDiagnosekoderFetch).toHaveReturned())
        expect(mockedSearchDiagnosekoderFetch).toHaveBeenCalledWith(inputCode, 1, (new AbortController()).signal)
        expect(rendered.getByText(fakeDiagnosekoder[0].code)).toBeInTheDocument()
        expect(rendered.getByText(fakeDiagnosekoder[0].text)).toBeInTheDocument()
    })

    test("clicking a diagnose row should select it", async () => {
        const onSelected = jest.fn((dk: Diagnosekode) => void 0);
        const rendered = render(<DiagnosekodeSearch onSelectedDiagnose={onSelected} />)
        mockedSearchDiagnosekoderFetch.mockReset().mockResolvedValue({
            diagnosekoder: fakeDiagnosekoder.slice(0,1),
            hasMore: false,
            pageNumber: 1,
        })
        await waitFor(() => expect(mockedSearchDiagnosekoderFetch).toHaveReturned())
        const diagnosekode1 = fakeDiagnosekoder[0];
        await userEvent.click(rendered.getByText(diagnosekode1.code))
        expect(onSelected).toHaveBeenCalledWith(diagnosekode1)

        onSelected.mockReset();
        const selectBtn = rendered.getByLabelText("Velg", {selector: "button"});
        await userEvent.click(selectBtn)
        expect(onSelected).toHaveBeenCalledWith(diagnosekode1)
    })
})