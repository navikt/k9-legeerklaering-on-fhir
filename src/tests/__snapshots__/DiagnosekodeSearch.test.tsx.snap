// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagnosekodeSearch works initial empty search returns pages of results 1`] = `
<table
  class="navds-table navds-table--small"
>
  <thead
    class="navds-table__header"
  >
    <tr
      class="navds-table__row navds-table__row--shade-on-hover"
    >
      <th
        class="navds-table__header-cell navds-label diagnosekodeCol"
        scope="col"
      >
        Kode
      </th>
      <th
        class="navds-table__header-cell navds-label"
        scope="col"
      >
        Beskrivelse
      </th>
      <th
        class="navds-table__header-cell navds-label actionCol"
        scope="col"
      />
    </tr>
  </thead>
  <tbody
    class="navds-table__body nonTransparent"
  >
    <tr
      class="navds-table__row diagnosekoderow navds-table__row--shade-on-hover"
    >
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        A000
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        Diagnose 1
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        <button
          aria-label="Velg"
          class="navds-button navds-button--tertiary navds-button--small navds-button--icon-only"
        >
          <span
            class="navds-button__icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14.03 4.97a.75.75 0 1 0-1.06 1.06l5.22 5.22H3.5a.75.75 0 0 0 0 1.5h14.69l-5.22 5.22a.75.75 0 1 0 1.06 1.06l6.5-6.5a.75.75 0 0 0 .22-.493m-.204-.55a.828.828 0 0 0-.016-.017l-6.5-6.5"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </td>
    </tr>
    <tr
      class="navds-table__row diagnosekoderow navds-table__row--shade-on-hover"
    >
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        A001
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        Diagnose 2
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        <button
          aria-label="Velg"
          class="navds-button navds-button--tertiary navds-button--small navds-button--icon-only"
        >
          <span
            class="navds-button__icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14.03 4.97a.75.75 0 1 0-1.06 1.06l5.22 5.22H3.5a.75.75 0 0 0 0 1.5h14.69l-5.22 5.22a.75.75 0 1 0 1.06 1.06l6.5-6.5a.75.75 0 0 0 .22-.493m-.204-.55a.828.828 0 0 0-.016-.017l-6.5-6.5"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </td>
    </tr>
    <tr
      class="navds-table__row diagnosekoderow navds-table__row--shade-on-hover"
    >
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        A002
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        Diagnose 3
      </td>
      <td
        class="navds-table__data-cell navds-body-short navds-body-short--medium"
      >
        <button
          aria-label="Velg"
          class="navds-button navds-button--tertiary navds-button--small navds-button--icon-only"
        >
          <span
            class="navds-button__icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14.03 4.97a.75.75 0 1 0-1.06 1.06l5.22 5.22H3.5a.75.75 0 0 0 0 1.5h14.69l-5.22 5.22a.75.75 0 1 0 1.06 1.06l6.5-6.5a.75.75 0 0 0 .22-.493m-.204-.55a.828.828 0 0 0-.016-.017l-6.5-6.5"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
